src/metadata/trackmetadatataglib.cpp

```diff
@@ -0,0 +1,768 @@
+    pTag->setAlbum(toTagLibString(trackMetadata.getAlbum()));
+    pTag->setGenre(toTagLibString(trackMetadata.getGenre()));
+    pTag->setComment(toTagLibString(trackMetadata.getComment()));
+
+    // Only write the year into the common tag if the corresponding
+    // field contains a 4-character numeric string.
+    const QString yearString(trackMetadata.getYear().trimmed());
+    if (4 == yearString.length()) {
+        bool yearValid = false;
+        uint year = trackMetadata.getYear().toUInt(&yearValid);
+        if (yearValid && (year > 0)) {
+            pTag->setYear(year);
+        }
+    }
+    // Derived tags might be able to write the complete string
+    // from trackMetadata.getYear() into the corresponding field.
+
+    bool trackNumberValid = false;
+    uint track = trackMetadata.getTrackNumber().toUInt(&trackNumberValid);
+    if (trackNumberValid && (track > 0)) {
+        pTag->setTrack(track);
+    }
+
+    return true;
+}
+
+bool writeID3v2Tag(TagLib::ID3v2::Tag* pTag,
+        const TrackMetadata& trackMetadata) {
+    if (NULL == pTag) {
+        return false;
+    }
+    const TagLib::ID3v2::Header* pHeader = pTag->header();
+    if ((NULL == pHeader) || (3 > pHeader->majorVersion())) {
+        // only ID3v2.3.x and higher (currently only ID3v2.4.x) are supported
+        return false;
+    }
+
+    // Write common metadata
+    writeTag(pTag, trackMetadata);
+
+    // additional tags
+    writeID3v2TextIdentificationFrame(pTag, "TPE2",
+            trackMetadata.getAlbumArtist());
+    // According to the specification "The 'TBPM' frame contains the number
+    // of beats per minute in the mainpart of the audio. The BPM is an integer
+    // and represented as a numerical string."
+    // Reference: http://id3.org/id3v2.3.0
+    writeID3v2TextIdentificationFrame(pTag, "TBPM",
+            TrackMetadata::formatBpm(trackMetadata.getBpmAsInteger()));
+    writeID3v2TextIdentificationFrame(pTag, "TKEY", trackMetadata.getKey());
+    writeID3v2TextIdentificationFrame(pTag, "TCOM",
+            trackMetadata.getComposer());
+    writeID3v2TextIdentificationFrame(pTag, "TIT1",
+            trackMetadata.getGrouping());
+    if (4 <= pHeader->majorVersion()) {
+        // ID3v2.4.0: TDRC replaces TYER + TDAT
+        writeID3v2TextIdentificationFrame(pTag, "TDRC",
+                trackMetadata.getYear());
+    } else {
+        // Fallback: Write TYER and TDAT
+        const QStringList yearParts(trackMetadata.getYear().split('-'));
+        if (0 < yearParts.length()) {
+            const QString year(yearParts[0].trimmed());
+            if (4 == year.length()) {
+                writeID3v2TextIdentificationFrame(pTag, "TYER",
+                        year); // yyyy
+                if (3 == yearParts.length()) {
+                    const QString month(yearParts[1].trimmed());
+                    const QString day(yearParts[2].trimmed());
+                    if ((2 == month.length()) && (2 == day.length())) {
```

Comment:
2 == day.length() 
will fail in the unlikely case that we have Time info, separated by " " or "T"